Building an offchain::ipfs app (#9)
rs-ipfs/offchain-ipfs-manual: src/template-pallet.md

# Example Pallet

`offchain::ipfs` comes with a showcase **[pallet]**, which is essentially a Rust module that
complies with the requirements to be included within a substrate **runtime**.

> > **This pallet is meant only as an example.** We include to be helpful for future pallet authors

Reviewer: ```suggestion
> **This pallet is meant only as an example.** We're including it to be helpful for future pallet authors
```

> that want to use the native runtime IPFS node to suit their needs.

Reviewer: ```suggestion
that want to use the embedded native IPFS node to suit their needs.
```

## Prelude

We start by using items from the native runtime. Our pallet is `no_std` since we're targeting `Wasm`

```rust,ignore
#![cfg_attr(not(feature = "std"), no_std)]
// ...
use sp_core::offchain::{
  Duration, IpfsRequest, IpfsResponse, OpaqueMultiaddr, Timestamp
};
// ...
use sp_runtime::offchain::ipfs;
```

## Command Types

When your JSON-RPC calls are received by the pallet, the requests are
expressed as `___Command` enums and stored in the off-chain worker storage as
a queue, to be ingested by your native runtime and passed to IPFS.

Derive attributes are omitted.

```rust,ignore
// Commands involved in peer-to-peer connections
enum ConnectionCommand {
    ConnectTo(OpaqueMultiaddr),
    DisconnectFrom(OpaqueMultiaddr),
}

// Commands that add, remove, pin, unpin, and output data
enum DataCommand {
    AddBytes(Vec<u8>),
    CatBytes(Vec<u8>),
    InsertPin(Vec<u8>),
    RemoveBlock(Vec<u8>),
    RemovePin(Vec<u8>),
}

// Commands that query the distributed hash table (DHT)
// for peers and content
enum DhtCommand {
    FindPeer(Vec<u8>),
    GetProviders(Vec<u8>),
}
```

## The runtime configuration trait

> The `system::Trait` trait (not to be confused with the Rust `trait` keyword), allows you to define

Reviewer: maybe link to the substrate tutorial on creating pallets or sth like that?

pub trait Trait: system::Trait { // Use traits here to tightly couple to runtime
    /// The overarching event type.
    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
}
```

Later in the code, we implement some helper functions on this trait. The function bodies are
omitted for brevity's sake.

```rust,ignore
impl<T: Trait> Module<T> {
    // "Sends" a request to the local IPFS node by adding it to the offchain storage
    fn ipfs_request(req: IpfsRequest, deadline: impl Into<Option<Timestamp>>)
      -> Result<IpfsResponse, Error<T>>

    // Reads from the `ConnectionQueue` and connects / disconnects
    // from desired / undesired peers, respectively
    fn connection_housekeeping() -> Result<(), Error<T>>

    // Reads `FindPeer` and `GetProviders` commands from the `DhtQueue`,
    // and requests their execution from the native runtime
    fn handle_dht_requests() -> Result<(), Error<T>>

    // Reads `AddBytes`, `CatBytes`, `DataCommand`, `RemoveBlock`, `InsertPin`,
    // and `RemovePin` commands from the `DataQueue` and requests their
    // execution from the native runtime.
    fn handle_data_requests() -> Result<(), Error<T>>

>     // Logs metadata (the number of connected peers) to the console at the DEBUG loglevel

Reviewer: ```suggestion
    // Logs metadata (the number of connected peers) to the console at the DEBUG log level
```

- `decl_module!`
- `decl_event!`
- `decl_storage!`
- `decl_error!`

[FRAME]: https://substrate.dev/docs/en/knowledgebase/runtime/frame

### `decl_storage!`

Here, we define the data that will actually be stored on-chain as **extrinsics**.

Since the offchain-worker can't perform I/O outside of the wasm context, we store our requests as
queues, to be processed on a periodic basis, consumed, and ultimately performed by the native
runtime.

This is where we use the `ConnectionQueue`, `DataQueue`, and `DhtQueue` command types from above.

```rust,ignore
// This pallet's storage items.
decl_storage! {
    trait Store for Module<T: Trait> as TemplateModule {
        // A list of addresses to connect to and disconnect from.
        pub ConnectionQueue: Vec<ConnectionCommand>;
        // A queue of data to publish or obtain on IPFS.
        pub DataQueue: Vec<DataCommand>;
        // A list of requests to the DHT.
        pub DhtQueue: Vec<DhtCommand>;
    }
}
```

### `decl_event!`

Once a command from the queue is processed by both the off-chain worker and the native runtime,
the IPFS response will be emitted as a chain event for anybody to consume as they need. This is

> where we define what those events are, when they happen, and what information they contain.

Author: Not sure if this is accurate since I'm a bit fuzzy on these events vs chain events

Reviewer: no, only the requests are emitted as events atm; the results could probably be registered as chain events, but it would seriously complicate the already-quite-complex showcase pallet as, if I remember correctly, the offchain worker would need to be able to emit signed transactions in order to be able to produce those

Author: Updated this section

}

// Commands that add, remove, pin, unpin, and output data
enum DataCommand {
    AddBytes(Vec<u8>),
    CatBytes(Vec<u8>),
    InsertPin(Vec<u8>),
    RemoveBlock(Vec<u8>),
    RemovePin(Vec<u8>),
}

// Commands that query the distributed hash table (DHT)
// for peers and content
enum DhtCommand {
    FindPeer(Vec<u8>),
    GetProviders(Vec<u8>),
}
```

## The runtime configuration trait

The `system::Trait` trait (not to be confused with the Rust `trait` keyword), allows you to define
which capabilities from the runtime you want to include, and how you want to use them. You can also
"tightly couple" your pallet to other pallets by adding their `Trait`s to your pallet's
inherited trait list.

Here, however, we keep things simple by:

1. Loosely coupling this pallet by leaving out inherited traits
2. Including only the required `Event` type

```rust,ignore
/// The pallet's configuration trait.
pub trait Trait: system::Trait { // Use traits here to tightly couple to runtime
    /// The overarching event type.
    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
}
```

Later in the code, we implement some helper functions on this trait. The function bodies are

> omitted for brevity's sake.

Reviewer 2: These helper functions are implemented for the Module struct instead of the trait Trait

        pub ConnectionQueue: Vec<ConnectionCommand>;
        // A queue of data to publish or obtain on IPFS.
        pub DataQueue: Vec<DataCommand>;
        // A list of requests to the DHT.
        pub DhtQueue: Vec<DhtCommand>;
    }
}
```

### `decl_event!`

This is where we define what those events are, when they happen, and what they contain.

Once a command is received by the off-chain worker, the corresponding handler from the `decl_module!`
section will "deposit" the event in its storage, to be processed by the native runtime.

```rust,ignore
// The pallet's events
decl_event!(
    pub enum Event<T> where AccountId = <T as system::Trait>::AccountId {
        ConnectionRequested(AccountId),
        DisconnectRequested(AccountId),
        QueuedDataToAdd(AccountId),
        QueuedDataToCat(AccountId),
        QueuedDataToPin(AccountId),
        QueuedDataToRemove(AccountId),
        QueuedDataToUnpin(AccountId),
        FindPeerIssued(AccountId),
        FindProvidersIssued(AccountId),
    }
);
```

### `decl_module!`

This section, perhaps the most critical section of any given pallet, is where you can define

> functions that are exposed as via JSON-RPC to client libraries and, by proxy, your users.

Reviewer 2: ```suggestion
functions that are exposed via JSON-RPC to client libraries and, by proxy, your users.
```

that want to use the embedded native IPFS node to suit their needs.

If you're familiar with Substrate and the Framework for Runtime Aggregation of Modularized Entities
(FRAME), you can simply view the [source code] for this pallet. Otherwise, read on as we go through
the code step by step.

Please note the order in which these concepts are explained here is not necessarily the order that
they appear in the code.

You can also learn more by following the [Building a Custom Pallet] tutorial.

[Building a Custom Pallet]: https://substrate.dev/docs/en/tutorials/build-a-dapp/pallet
[source code]: https://github.com/rs-ipfs/substrate/blob/offchain_ipfs_bleeding_edge/bin/node-template/pallets/template/src/lib.rs
[pallet]: https://substrate.dev/docs/en/knowledgebase/runtime/pallets

## Prelude

We start by using items from the native runtime. Our pallet is `no_std` since we're targeting `Wasm`

```rust,ignore
#![cfg_attr(not(feature = "std"), no_std)]
// ...
use sp_core::offchain::{
  Duration, IpfsRequest, IpfsResponse, OpaqueMultiaddr, Timestamp
};
// ...
use sp_runtime::offchain::ipfs;
```

## Command Types

When your JSON-RPC calls are received by the pallet, the requests are

> expressed as `___Command` enums and stored in the off-chain worker storage as

Reviewer: ```suggestion
expressed as `___Command` enums and stored in the offchain worker storage as
```
for consistence with the rest of the content

    fn ipfs_request(req: IpfsRequest, deadline: impl Into<Option<Timestamp>>)
      -> Result<IpfsResponse, Error<T>>

    // Reads from the `ConnectionQueue` and connects / disconnects
    // from desired / undesired peers, respectively
    fn connection_housekeeping() -> Result<(), Error<T>>

    // Reads `FindPeer` and `GetProviders` commands from the `DhtQueue`,
    // and requests their execution from the native runtime
    fn handle_dht_requests() -> Result<(), Error<T>>

    // Reads `AddBytes`, `CatBytes`, `DataCommand`, `RemoveBlock`, `InsertPin`,
    // and `RemovePin` commands from the `DataQueue` and requests their
    // execution from the native runtime.
    fn handle_data_requests() -> Result<(), Error<T>>

    // Logs metadata (the number of connected peers) to the console at the DEBUG log level
    fn print_metadata() -> Result<(), Error<T>>
```

## The `decl_` macros

Pallets included in Substrate runtimes must adhere to the conventions of [FRAME].
In practice, this means you must implement `decl_` macros:

- `decl_module!`
- `decl_event!`
- `decl_storage!`
- `decl_error!`

[FRAME]: https://substrate.dev/docs/en/knowledgebase/runtime/frame

> 

Reviewer: ```suggestion
```

### `decl_storage!`

> Here, we define the data that will actually be stored on-chain as **extrinsics**.

Reviewer: ```suggestion
Here, we define the data that will actually be stored on-chain when calling **extrinsics**.
```

Since the offchain-worker can't perform I/O outside of the wasm context, we store our requests as
queues, to be processed on a periodic basis, consumed, and ultimately performed by the native
runtime.

This is where we use the `ConnectionQueue`, `DataQueue`, and `DhtQueue` command types from above.

```rust,ignore
// This pallet's storage items.
decl_storage! {
    trait Store for Module<T: Trait> as TemplateModule {
        // A list of addresses to connect to and disconnect from.
        pub ConnectionQueue: Vec<ConnectionCommand>;
        // A queue of data to publish or obtain on IPFS.
        pub DataQueue: Vec<DataCommand>;
        // A list of requests to the DHT.
        pub DhtQueue: Vec<DhtCommand>;
    }
}
```

### `decl_event!`

> This is where we define what those events are, when they happen, and what they contain.

Reviewer: ```suggestion
This is where we define what those events are, and what they contain.
```
the call sites determine when they occur

> Once a command is received by the off-chain worker, the corresponding handler from the `decl_module!`

Reviewer: ```suggestion
Once a command is sent to the offchain worker, one of the following chain events is emitted.
```

> section will "deposit" the event in its storage, to be processed by the native runtime.

Reviewer: ```suggestion
```

```rust,ignore
// The pallet's events
decl_event!(
    pub enum Event<T> where AccountId = <T as system::Trait>::AccountId {
        ConnectionRequested(AccountId),
        DisconnectRequested(AccountId),
        QueuedDataToAdd(AccountId),
        QueuedDataToCat(AccountId),
        QueuedDataToPin(AccountId),
        QueuedDataToRemove(AccountId),
        QueuedDataToUnpin(AccountId),
        FindPeerIssued(AccountId),
        FindProvidersIssued(AccountId),
    }
);
```

### `decl_module!`

This section, perhaps the most critical section of any given pallet, is where you can define
functions that are exposed as via JSON-RPC to client libraries and, by proxy, your users.

> In practice, the bulk of what these functions do is to modify the `DataQueue`, `DhtQueue`, and

Reviewer: ```suggestion
In practice, the bulk of what these functions do is modify the `DataQueue`, `DhtQueue`, and
```

`ConnectionQueue` storage objects by pushing signed command requests to their respective queues.

> Some default types functions are omitted, but we've kept the `#[weight]` attributes around.

Reviewer: ```suggestion
Some default types in the functions are omitted, but we've kept the `#[weight]` attributes around.
```

> > The Substrate docs define defines one unit of weight as "one picosecond of execution time on

Reviewer: ```suggestion
> The Substrate docs define one unit of weight as "one picosecond of execution time on
```

> fixed reference hardware." These are essentially _time limits_ for block creation, and can be

Reviewer: ```suggestion
fixed reference hardware". These are essentially _time limits_ for block creation, and can be
```

> (indirectly) mapped to transaction fees analogous to something like "gas fees." Read more about

Reviewer: ```suggestion
(indirectly) mapped to transaction fees analogous to something like "gas fees". Read more about
```
